panic: `index out of range` when pressing "enter" without a default sub set

Repro:
1. Run `azd config unset default.subscription`
1. In new repository, run `azd pipeline config` (should happen with other commands as well) that triggers the new environment creation flow
1. When prompted "Please select an Azure Subscription to use", it seems that the first entry is highlighted. Press enter.

Expected:
The first entry shouldn't be highlighted if it is not selected. If it is selected, I would expect the command to succeed.

Actual:
Observed panic (with debug log + print out of default)

```
2022/11/01 21:24:33 util.go:234: subscriptionOptions: [ 1. Visual Studio Enterprise with MSDN (7c5b731d-a3c1-43b6-8aab-a8eb3e5fc173) Other (enter manually)], defaultSubscription: 
? Please select an Azure Subscription to use: 
2022/11/01 21:24:33 util.go:241: subscriptionSelectionIndex: -1, err: <nil>
panic: runtime error: index out of range [-1]

goroutine 1 [running]:
github.com/azure/azure-dev/cli/azd/cmd.ensureEnvironmentInitialized({0xbfabb0, 0xc000381d10}, {{0xc00048a950, 0x9}, {0x0, 0x0}, {0x0, 0x0}}, 0xc0004826d8, {0xbfbff8, ...})
        D:/repos/azure-dev/cli/azd/cmd/util.go:247 +0xb52
github.com/azure/azure-dev/cli/azd/cmd.loadOrInitEnvironment({0xbfabb0, 0xc000381d10}, 0xc0001b5d10, 0xc000338be0, {0xbfbff8, 0xc000363f80})
        D:/repos/azure-dev/cli/azd/cmd/util.go:168 +0x1a5
github.com/azure/azure-dev/cli/azd/cmd.(*pipelineConfigAction).Run(0xc000363f10, {0xbfabb0, 0xc000381d10})
        D:/repos/azure-dev/cli/azd/cmd/pipeline.go:113 +0x1ae
github.com/azure/azure-dev/cli/azd/cmd.BuildCmd[...].func1({0xbfabb0?, 0xc000381aa0}, {0xc000381a40, 0x0?, 0x0?})
        D:/repos/azure-dev/cli/azd/cmd/root.go:155 +0x154
github.com/azure/azure-dev/cli/azd/cmd.BuildCmd[...].func2.1()
        D:/repos/azure-dev/cli/azd/cmd/root.go:163 +0x44
github.com/azure/azure-dev/cli/azd/cmd.runCmdWithTelemetry(0xc000386000, 0xc0002bfa48)
        D:/repos/azure-dev/cli/azd/cmd/root.go:177 +0x14e
github.com/azure/azure-dev/cli/azd/cmd.BuildCmd[...].func2({0xc000381a40?, 0x3?, 0x3?})
        D:/repos/azure-dev/cli/azd/cmd/root.go:164 +0xb3
github.com/spf13/cobra.(*Command).execute(0xc000386000, {0xc000381a10, 0x3, 0x3})
        C:/Users/weilim/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:856 +0x67c
github.com/spf13/cobra.(*Command).ExecuteC(0xc000356f00)
        C:/Users/weilim/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:974 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
        C:/Users/weilim/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:902
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        C:/Users/weilim/go/pkg/mod/github.com/spf13/cobra@v1.3.0/command.go:895
```

--- a/cli/azd/cmd/util.go
+++ b/cli/azd/cmd/util.go
@@ -285,7 +285,7 @@ func ensureEnvironmentInitialized(
 }
 
 func getSubscriptionOptions(ctx context.Context) ([]string, any, error) {
-	accountManager, err := account.NewManager(config.NewManager(), azcli.GetAzCli(ctx))
+	accountManager, err := account.NewManager(config.GetConfigManager(ctx), azcli.GetAzCli(ctx))
 	if err != nil {
 		return nil, nil, fmt.Errorf("failed creating account manager: %w", err)
 	}

--- a/cli/azd/cmd/util.go
+++ b/cli/azd/cmd/util.go
@@ -284,15 +284,15 @@ func ensureEnvironmentInitialized(
 	return nil
 }
 
-func getSubscriptionOptions(ctx context.Context) ([]string, string, error) {
+func getSubscriptionOptions(ctx context.Context) ([]string, any, error) {
 	accountManager, err := account.NewManager(config.NewManager(), azcli.GetAzCli(ctx))
 	if err != nil {
-		return nil, "", fmt.Errorf("failed creating account manager: %w", err)
+		return nil, nil, fmt.Errorf("failed creating account manager: %w", err)
 	}
 
 	subscriptionInfos, err := accountManager.GetSubscriptions(ctx)
 	if err != nil {
-		return nil, "", fmt.Errorf("listing accounts: %w", err)
+		return nil, nil, fmt.Errorf("listing accounts: %w", err)
 	}
 
 	// If `AZURE_SUBSCRIPTION_ID` is set in the environment, use it to influence
@@ -308,7 +308,7 @@ func getSubscriptionOptions(ctx context.Context) ([]string, string, error) {
 	}
 
 	var subscriptionOptions = make([]string, len(subscriptionInfos)+1)
-	var defaultSubscription string = ""
+	var defaultSubscription any
 
 	for index, info := range subscriptionInfos {
 		subscriptionOptions[index] = fmt.Sprintf("%2d. %s (%s)", index+1, info.Name, info.Id)

--- a/cli/azd/cmd/util_test.go
+++ b/cli/azd/cmd/util_test.go
@@ -12,6 +12,7 @@ import (
 	"testing"
 
 	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armsubscriptions"
 	"github.com/azure/azure-dev/cli/azd/pkg/convert"
 	"github.com/azure/azure-dev/cli/azd/pkg/infra"
 	"github.com/azure/azure-dev/cli/azd/pkg/input"
@@ -109,5 +110,80 @@ func Test_promptEnvironmentName(t *testing.T) {
 
 		sort.Strings(groups)
 		require.Equal(t, []string{"groupA", "groupB", "groupC"}, groups)
+	})
+}
+
+func Test_getSubscriptionOptions(t *testing.T) {
+	t.Run("no default", func(t *testing.T) {
+		mockContext := mocks.NewMockContext(context.Background())
+		mockContext.HttpClient.When(func(request *http.Request) bool {
+			return strings.Contains(request.URL.Path, "subscriptions/")
+		}).RespondFn(func(request *http.Request) (*http.Response, error) {
+			return mocks.CreateHttpResponseWithBody(request, 200, armsubscriptions.Subscription{
+				ID:             convert.RefOf("SUBSCRIPTION"),
+				SubscriptionID: convert.RefOf("SUBSCRIPTION_ID"),
+				DisplayName:    convert.RefOf("DISPLAY"),
+				TenantID:       convert.RefOf("TENANT"),
+			})
+		})
+		mockContext.HttpClient.When(func(request *http.Request) bool {
+			return request.URL.Path == "/subscriptions"
+		}).RespondFn(func(request *http.Request) (*http.Response, error) {
+			return mocks.CreateHttpResponseWithBody(request, 200, armsubscriptions.ClientListResponse{
+				SubscriptionListResult: armsubscriptions.SubscriptionListResult{
+					Value: []*armsubscriptions.Subscription{
+						{
+							ID:             convert.RefOf("SUBSCRIPTION"),
+							SubscriptionID: convert.RefOf("SUBSCRIPTION_ID_OTHER"),
+							DisplayName:    convert.RefOf("DISPLAY"),
+							TenantID:       convert.RefOf("TENANT"),
+						},
+					},
+				},
+			})
+		})
+		subList, result, err := getSubscriptionOptions(*mockContext.Context)
+
+		require.Nil(t, err)
+		require.EqualValues(t, 2, len(subList))
+		require.EqualValues(t, nil, result)
+	})
+
+	t.Run("default", func(t *testing.T) {
+		mockContext := mocks.NewMockContext(context.Background())
+		mockContext.HttpClient.When(func(request *http.Request) bool {
+			return strings.Contains(request.URL.Path, "subscriptions/")
+		}).RespondFn(func(request *http.Request) (*http.Response, error) {
+			return mocks.CreateHttpResponseWithBody(request, 200, armsubscriptions.Subscription{
+				ID:             convert.RefOf("SUBSCRIPTION"),
+				SubscriptionID: convert.RefOf("SUBSCRIPTION_ID"),
+				DisplayName:    convert.RefOf("DISPLAY"),
+				TenantID:       convert.RefOf("TENANT"),
+			})
+		})
+		mockContext.HttpClient.When(func(request *http.Request) bool {
+			return request.URL.Path == "/subscriptions"
+		}).RespondFn(func(request *http.Request) (*http.Response, error) {
+			return mocks.CreateHttpResponseWithBody(request, 200, armsubscriptions.ClientListResponse{
+				SubscriptionListResult: armsubscriptions.SubscriptionListResult{
+					Value: []*armsubscriptions.Subscription{
+						{
+							ID:             convert.RefOf("SUBSCRIPTION"),
+							SubscriptionID: convert.RefOf("SUBSCRIPTION_ID"),
+							DisplayName:    convert.RefOf("DISPLAY"),
+							TenantID:       convert.RefOf("TENANT"),
+						},
+					},
+				},
+			})
+		})
+		subList, result, err := getSubscriptionOptions(*mockContext.Context)
+
+		require.Nil(t, err)
+		require.EqualValues(t, 2, len(subList))
+		require.NotNil(t, result)
+		defSub, ok := result.(string)
+		require.True(t, ok)
+		require.EqualValues(t, " 1. DISPLAY (SUBSCRIPTION_ID)", defSub)
 	})
 }

--- a/cli/azd/cmd/util_test.go
+++ b/cli/azd/cmd/util_test.go
@@ -13,6 +13,7 @@ import (
 
 	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
 	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armsubscriptions"
+	"github.com/azure/azure-dev/cli/azd/pkg/config"
 	"github.com/azure/azure-dev/cli/azd/pkg/convert"
 	"github.com/azure/azure-dev/cli/azd/pkg/infra"
 	"github.com/azure/azure-dev/cli/azd/pkg/input"
@@ -114,18 +115,10 @@ func Test_promptEnvironmentName(t *testing.T) {
 }
 
 func Test_getSubscriptionOptions(t *testing.T) {
-	t.Run("no default", func(t *testing.T) {
+	t.Run("no default config set", func(t *testing.T) {
 		mockContext := mocks.NewMockContext(context.Background())
-		mockContext.HttpClient.When(func(request *http.Request) bool {
-			return strings.Contains(request.URL.Path, "subscriptions/")
-		}).RespondFn(func(request *http.Request) (*http.Response, error) {
-			return mocks.CreateHttpResponseWithBody(request, 200, armsubscriptions.Subscription{
-				ID:             convert.RefOf("SUBSCRIPTION"),
-				SubscriptionID: convert.RefOf("SUBSCRIPTION_ID"),
-				DisplayName:    convert.RefOf("DISPLAY"),
-				TenantID:       convert.RefOf("TENANT"),
-			})
-		})
+		// set empty config as mock
+		mockContext.ConfigManager.WithConfig(config.NewConfig(nil))
 		mockContext.HttpClient.When(func(request *http.Request) bool {
 			return request.URL.Path == "/subscriptions"
 		}).RespondFn(func(request *http.Request) (*http.Response, error) {
@@ -149,10 +142,23 @@ func Test_getSubscriptionOptions(t *testing.T) {
 		require.EqualValues(t, nil, result)
 	})
 
-	t.Run("default", func(t *testing.T) {
+	t.Run("default value set", func(t *testing.T) {
+		// mocked config
+		configSubscriptionId := "theSubscriptionInTheConfig"
+		c := config.NewConfig(nil)
+		err := c.Set("defaults.location", "location")
+		require.Nil(t, err)
+		err = c.Set("defaults.subscription", configSubscriptionId)
+		require.Nil(t, err)
+
+		// mocks
 		mockContext := mocks.NewMockContext(context.Background())
+		mockContext.ConfigManager.WithConfig(c)
+
+		// Mock the account returned when a config is found
+		// the url path should contain the sub name from the config file
 		mockContext.HttpClient.When(func(request *http.Request) bool {
-			return strings.Contains(request.URL.Path, "subscriptions/")
+			return request.URL.Path == ("/subscriptions/" + configSubscriptionId)
 		}).RespondFn(func(request *http.Request) (*http.Response, error) {
 			return mocks.CreateHttpResponseWithBody(request, 200, armsubscriptions.Subscription{
 				ID:             convert.RefOf("SUBSCRIPTION"),
@@ -161,6 +167,8 @@ func Test_getSubscriptionOptions(t *testing.T) {
 				TenantID:       convert.RefOf("TENANT"),
 			})
 		})
+		// Mock other subscriptions for the user, as azd will merge
+		// the default account with all others accessible
 		mockContext.HttpClient.When(func(request *http.Request) bool {
 			return request.URL.Path == "/subscriptions"
 		}).RespondFn(func(request *http.Request) (*http.Response, error) {
@@ -177,6 +185,8 @@ func Test_getSubscriptionOptions(t *testing.T) {
 				},
 			})
 		})
+
+		// finally invoking the test
 		subList, result, err := getSubscriptionOptions(*mockContext.Context)
 
 		require.Nil(t, err)

--- a/cli/azd/pkg/config/manager.go
+++ b/cli/azd/pkg/config/manager.go
@@ -1,6 +1,7 @@
 package config
 
 import (
+	"context"
 	"encoding/json"
 	"fmt"
 	"io"
@@ -26,6 +27,24 @@ type Manager interface {
 // Creates a new Configuration Manager
 func NewManager() Manager {
 	return &manager{}
+}
+
+type contextKey string
+
+const (
+	configManagerContextKey contextKey = "configManagerContextKey"
+)
+
+func WithConfigManager(ctx context.Context, manager Manager) context.Context {
+	return context.WithValue(ctx, configManagerContextKey, manager)
+}
+
+func GetConfigManager(ctx context.Context) Manager {
+	configManager, ok := ctx.Value(configManagerContextKey).(Manager)
+	if ok {
+		return configManager
+	}
+	return NewManager()
 }
 
 // Saves the azd configuration to the specified file path

--- a/cli/azd/test/mocks/mock_context.go
+++ b/cli/azd/test/mocks/mock_context.go
@@ -5,6 +5,7 @@ import (
 	"os"
 
 	"github.com/azure/azure-dev/cli/azd/internal"
+	"github.com/azure/azure-dev/cli/azd/pkg/config"
 	"github.com/azure/azure-dev/cli/azd/pkg/exec"
 	"github.com/azure/azure-dev/cli/azd/pkg/httputil"
 	"github.com/azure/azure-dev/cli/azd/pkg/identity"
@@ -40,6 +41,7 @@ func NewMockContext(ctx context.Context) *MockContext {
 	ctx = httputil.WithHttpClient(ctx, httpClient)
 	ctx = identity.WithCredentials(ctx, &credentials)
 	ctx = output.WithWriter(ctx, os.Stdout)
+	ctx = config.WithConfigManager(ctx, configManager)
 
 	mockContext := &MockContext{
 		Context:       &ctx,

